distsql: support DISTINCT

rjnn opened the pull request:
Support DISTINCT in the DistSQL planner. Closes #14141.

Currently failing DistSQL logic tests perplexingly, as if I start up a local cluster, use `SET distsql = always` and run them manually, they pass.

Feedback welcome on the TODO, or a better way to address that shortcoming in planning without any statistics.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14300)
<!-- Reviewable:end -->

RaduBerinde commented:
> Currently failing DistSQL logic tests perplexingly
The logic tests sets up a fake span resolver which simulates random distributions of keys.

RaduBerinde commented:
You can set the env `COCKROACH_DISTSQL_LOG_PLAN=1` to see plans in logs for the statements that fail. The only thing I can think of is if the ordering that the processor depends on (through `OrderedColumns/columnsInOrder`) is not being maintained correctly.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlplan/physical_plan.go, line 631 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KfmzAqL9pQh6IQss-Rc:-KfmzAqL9pQh6IQss-Rd:b-x9iee3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0492e07d0f5ce31074f776ec081c1a7d96b6b361/pkg/sql/distsqlplan/physical_plan.go#L631)):*
> ```Go
> }
> 
> // AddDistinct adds distinct nodes to the results of the current plan.
> ```

This belongs in `distsql_physical_planner.go`; `physical_plan.go` has only more basic building-blocks.

---

*[pkg/sql/distsqlplan/physical_plan.go, line 646 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kfmzi0lxFgQVAkUeupD:-Kfmzi0lxFgQVAkUeupE:b-upl71w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0492e07d0f5ce31074f776ec081c1a7d96b6b361/pkg/sql/distsqlplan/physical_plan.go#L646)):*
> ```Go
> 	}
> 	// TODO(arjun): This is potentially inefficient, especially if we don't have any sorted columns.
> 	if len(currentResultRouters) > 1 {
> ```

if we have a single result router, we should do the opposite: add just the `NoGroupingStage` (which will stay on whatever node generates that stream).

---

*[pkg/sql/distsqlplan/physical_plan.go, line 651 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KfmzR8OzKXrSCv4bQ7F:-KfmzR8OzKXrSCv4bQ7G:bolhoa0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0492e07d0f5ce31074f776ec081c1a7d96b6b361/pkg/sql/distsqlplan/physical_plan.go#L651)):*
> ```Go
> 	}
> 
> 	p.AddSingleGroupStage(node, distinctSpec, distsqlrun.PostProcessSpec{}, p.ResultTypes)
> ```

Add a TODO here saying that we could distribute the final stage (by hash)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-Kfn00Rmbv64rTDQR9bQ:bl8xakp)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Thanks for the tip (and thank you for the review as well!). I will take a look at the logs to debug it, add the aggregation support and get back to you.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-KfnoHrPTGTQBHGdm-uX:bx6tfd2)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
RFAL due to fixing the distinct processor, which was broken on DISTINCTs who's inputs included columns that were to be excluded from the deduplication test. Passes logic tests, but does not yet support `SELECT SUM(DISTINCT)`. That is still WIP. Also rebased on master.

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlplan/physical_plan.go, line 631 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KfmzAqL9pQh6IQss-Rc:-KgFaezJEsIYs7INnY0G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0492e07d0f5ce31074f776ec081c1a7d96b6b361/pkg/sql/distsqlplan/physical_plan.go#L631)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This belongs in `distsql_physical_planner.go`; `physical_plan.go` has only more basic building-blocks.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlplan/physical_plan.go, line 646 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kfmzi0lxFgQVAkUeupD:-Kfr2e9y4dJM8E7MQjUI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0492e07d0f5ce31074f776ec081c1a7d96b6b361/pkg/sql/distsqlplan/physical_plan.go#L646)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

if we have a single result router, we should do the opposite: add just the `NoGroupingStage` (which will stay on whatever node generates that stream).
</blockquote></details>

Done.

---

*[pkg/sql/distsqlplan/physical_plan.go, line 651 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KfmzR8OzKXrSCv4bQ7F:-Kfr0eoy8_dvBaF7SeQm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0492e07d0f5ce31074f776ec081c1a7d96b6b361/pkg/sql/distsqlplan/physical_plan.go#L651)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Add a TODO here saying that we could distribute the final stage (by hash)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-KgGei0SlmHZh2grMjHG:bnd8qwh)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
It appears to me that this supports `SELECT SUM(DISTINCT)` just fine already. What more needs to be done? The following works:
```
INSERT INTO xyz VALUES (1, 2, 3), (2, 5, 6), (3, 2, 3), (4, 5, 6), (5, 2, 6), (6, 3, 5), (7, 2, 9);
set distsql = 'always';
SELECT sum (distinct y) FROM xyz;
+-----------------+
| sum(DISTINCT y) |
+-----------------+
|              10 |
+-----------------+
```

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-KgGmbew0GFEazGbuMoD:b88v7el)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Ah - I think what is missing is that with `SUM (DISTINCT y)` we don't do any local processing. This should be easily visible in the physical plan (`EXPLAIN (DISTSQL)`). We would want a local "distinct" step on each node, before the final stage.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-KgH50MUYCvsc-XD3Jvd:b-6f8br0)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
PTAL, I've added some work, on `SUM (DISTINCT y)`, as well as fixed the disabling of columns, it could use a look through. There's one outstanding TODO that I could get your feedback on as well.

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-KgQaeiKiWy7j1lUS7sg:b-7bqceq)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Nice work!

---

Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_physical_planner.go, line 802 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgR9zb8UvYyPFnKwNfY:-KgR9zb8UvYyPFnKwNfZ:b75jfcu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L802)):*
> ```Go
> 				// We can't do local distinct if we have a mix of distinct and
> 				// non-distinct aggregations.
> 				allDistinct = false
> ```

So it's clear how the local stage works if there is a single aggregation, or if they all have DISTINCT on the same column. However, the case where the columns differ is more subtle - e.g. `SELECT SUM(DISTINCT x), SUM(DISTINCT y)`.  In this case, we are saying that we can deduplicate across both `x` and `y` (so we allow all pairs of values, but no duplicates for any pair of values). This is pretty cool, but should be explained in a comment (perhaps below where we calculate distinctColumns).

---

*[pkg/sql/distsql_physical_planner.go, line 837 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgRC4uHFBU7Ha8blqIo:-KgRC4uHFBU7Ha8blqIp:b-edfn8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L837)):*
> ```Go
> 
> 		var orderedColumns []uint32
> 		// TODO(DONOTMERGE): Where do I get ordering information from?
> ```

From `n.plan` (the plan "under" the groupNode) - the existing physical plan corresponds to `n.plan` and should match that ordering (use `dsp.convertOrdering()` to convert to stream columns).

---

*[pkg/sql/distsql_physical_planner.go, line 865 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgRAWh8Lmde1B3ZFy5Q:-KgRAWh94mohZLajkU5Q:b-2s83ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L865)):*
> ```Go
> 			GroupCols:    groupCols,
> 		}
> 	} else if !multiStage {
> ```

[nit] this could be separated out from the other `if`, and we wouldn't need to duplicae the `finalAggSpec = ..` stuff

---

*[pkg/sql/distsql_physical_planner.go, line 1359 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgRClqLdYc0V5xRUjQe:-KgRClqLdYc0V5xRUjQf:bvuffqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L1359)):*
> ```Go
> 	}
> 
> 	plan.AddNoGroupingStage(distinctSpec, distsqlrun.PostProcessSpec{}, plan.ResultTypes, plan.MergeOrdering)
> ```

[nit] this case should be under the `if` (it's the special case, a reader that skims this shouldn't be deceived into thinking that we end in a "no grouping" stage in general).

---

*[pkg/sql/distsql_physical_planner_test.go, line 306 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgRDUD-cZFuie_Ghw1c:-KgRDUD-cZFuie_Ghw1d:b-ojeqkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner_test.go#L306)):*
> ```Go
> 			[][]string{{"5", "5"}},
> 		)
> 	})
> ```

can we add a few more tests with some ordered columns? Or just add a `TODO`, it will be easier once these are all logic tests.

---

*[pkg/sql/distsqlrun/distinct_test.go, line 85 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgREGE99fc165W4xoJy:-KgREGE99fc165W4xoJz:btlvp6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsqlrun/distinct_test.go#L85)):*
> ```Go
> 			},
> 		},
> 	}
> ```

We should add a couple of tests with fewer DistinctColumns

---

*[pkg/sql/distsqlrun/processors.proto, line 205 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgRDvtp4xzNa4hFqNVK:-KgRDvtp4xzNa4hFqNVL:b-9rg6dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsqlrun/processors.proto#L205)):*
> ```protobuf
>   // are specified.
>   repeated uint32 ordered_columns = 1;
>   repeated uint32 distinct_columns = 2;
> ```

This needs documentation (it's not trivial what the semantics are). One way of saying it is: if A,B,C are distinct_columns, for any possible set of A,B,C values we only output one (the first) row with those A,B,C values

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-KgREiph2kB7zjoZqBCb:bdtx8r6)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
TFTR! Failing a couple tests, which I will take a look at shortly, which should result in a couple fixes, but I think I've covered the intents of your comments so far.

---

Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/distsql_physical_planner.go, line 802 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgR9zb8UvYyPFnKwNfY:-KgW4fCbUlXwTxfgKaH7:b-1dqspu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L802)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

So it's clear how the local stage works if there is a single aggregation, or if they all have DISTINCT on the same column. However, the case where the columns differ is more subtle - e.g. `SELECT SUM(DISTINCT x), SUM(DISTINCT y)`.  In this case, we are saying that we can deduplicate across both `x` and `y` (so we allow all pairs of values, but no duplicates for any pair of values). This is pretty cool, but should be explained in a comment (perhaps below where we calculate distinctColumns).
</blockquote></details>

Is that how it works? `SELECT SUM(DISTINCT x), SUM(DISTINCT y)` treats the two cols independently, deduplicates the two cols independently (so a deduplicated x and a deduplicated y) and then sums them.
The behavior you're describing in prose would come from `SELECT sum(y), sum(z) from (select DISTINCT y, z FROM xyz);`

---

*[pkg/sql/distsql_physical_planner.go, line 837 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgRC4uHFBU7Ha8blqIo:-KgUvGdecisWPQGsenhH:b-gk5vqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L837)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

From `n.plan` (the plan "under" the groupNode) - the existing physical plan corresponds to `n.plan` and should match that ordering (use `dsp.convertOrdering()` to convert to stream columns).
</blockquote></details>

Done, I think.

---

*[pkg/sql/distsql_physical_planner.go, line 865 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgRAWh8Lmde1B3ZFy5Q:-KgVi6dr8AvnES1INMfo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L865)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] this could be separated out from the other `if`, and we wouldn't need to duplicae the `finalAggSpec = ..` stuff
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 1359 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgRClqLdYc0V5xRUjQe:-KgVjH43broDxhm6LV6x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L1359)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] this case should be under the `if` (it's the special case, a reader that skims this shouldn't be deceived into thinking that we end in a "no grouping" stage in general).
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner_test.go, line 306 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgRDUD-cZFuie_Ghw1c:-KgVkMJhOErfoEaV0Rxj:bxkwxyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner_test.go#L306)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

can we add a few more tests with some ordered columns? Or just add a `TODO`, it will be easier once these are all logic tests.
</blockquote></details>

Done. (Added the todo)

---

*[pkg/sql/distsqlrun/distinct_test.go, line 85 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgREGE99fc165W4xoJy:-KgVofNqpckIJb-lWcwt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsqlrun/distinct_test.go#L85)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should add a couple of tests with fewer DistinctColumns
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/processors.proto, line 205 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgRDvtp4xzNa4hFqNVK:-KgW3UJIGmOQ9AvsFGgr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsqlrun/processors.proto#L205)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This needs documentation (it's not trivial what the semantics are). One way of saying it is: if A,B,C are distinct_columns, for any possible set of A,B,C values we only output one (the first) row with those A,B,C values
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-KgW5SMwc-wl_hh_MCXh:b-kvka3)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_physical_planner.go, line 802 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgR9zb8UvYyPFnKwNfY:-KgWV6-j82EFKUcVTFTT:bffyt96) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L802)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Is that how it works? `SELECT SUM(DISTINCT x), SUM(DISTINCT y)` treats the two cols independently, deduplicates the two cols independently (so a deduplicated x and a deduplicated y) and then sums them.
The behavior you're describing in prose would come from `SELECT sum(y), sum(z) from (select DISTINCT y, z FROM xyz);`
</blockquote></details>

Yes, of course, that's what the statement does, and what the final stage does. I am talking about the local distinct stage that you are adding. What does the local stage do in this case?

---

*[pkg/sql/distsql_physical_planner.go, line 841 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgWVjTcXU1LRcls1NCg:-KgWVjTcXU1LRcls1NCh:blnl67e) ([raw file](https://github.com/cockroachdb/cockroach/blob/79ac1bb16b9a7fd0cc8473bdcf7ecd2aefe5b926/pkg/sql/distsql_physical_planner.go#L841)):*
> ```Go
> 		// to reduce streaming duplicates, and aggregate on the final node.
> 
> 		orderings := dsp.convertOrdering(n.plan.Ordering().ordering, p.planToStreamColMap).Columns
> ```

[nit] should be ordering

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300)*
<!-- Sent from Reviewable.io -->

rjnn committed:
distsql: fix distinct processor semantics

The distinct processor needs a filter on which columns to consider
when checking equality across rows. This enables columns to be passed
through the distinct processor. Fixes #14403.

rjnn committed:
distsql: add planning support for DISTINCT

Add planning support and associated tests for SELECT DISTINCT. This
does not add planning support for DISTINCT when it is part of an
aggregation, only when used as a standalone filter.

arjunravinarayan commented:
Refactored and cleaned up into three separate commits. I have one spot that that needs fixing, as I've got the `intermediateTypes` wrong (in `r6`), and would like you to take a look. I will continue plugging away at debugging that, but if you have any input that would be appreciated.

---

Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsql_physical_planner.go, line 802 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgR9zb8UvYyPFnKwNfY:-KgfXXTJNG6OlDu9R35X:byqhbkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L802)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yes, of course, that's what the statement does, and what the final stage does. I am talking about the local distinct stage that you are adding. What does the local stage do in this case?
</blockquote></details>

Sorry, I was misunderstanding your description in prose. I read your prose as describing a different behavior than the SQL you stated. I believe this code does the right thing in either case. The first example (`SELECT SUM(DISTINCT x), SUM(DISTINCT y)`) would have two separate aggregations, and each would have its own distinct stage. Those distinct stages obey the 'allDistinct` clause, so there would be local deduplication, and then a final deduplication before aggregation. The second example `SELECT DISTINCT y,z` would have a distinct processor (a regular one created by `createPlanForDistinct`) that fed into two separate aggregations, and that distinct processor can have local stages as well.

---

*[pkg/sql/distsql_physical_planner.go, line 841 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgWVjTcXU1LRcls1NCg:-KgaD66H0L7AuAJO4m1p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/79ac1bb16b9a7fd0cc8473bdcf7ecd2aefe5b926/pkg/sql/distsql_physical_planner.go#L841)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] should be ordering
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 850 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgfsT02hu-KsyYHuBlM:-KgfsT036LEyccmNkZUN:bz3oleb) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f7780419b953eb2fc769c53857415874d6d2dc/pkg/sql/distsql_physical_planner.go#L850)):*
> ```Go
> 		}
> 
> 		var intermediateTypes []sqlbase.ColumnType
> ```

This is almost certainly wrong, and I'm messing up setting it correctly. If you have any thoughts, I'd greatly appreciate it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-Kgfs8veYHSqunqB-ZGi:butmo3i)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_physical_planner.go, line 802 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgR9zb8UvYyPFnKwNfY:-KgftC0u1qD-CjjtAgB_:b-6qe2ae) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L802)):*
> The first example (SELECT SUM(DISTINCT x), SUM(DISTINCT y)) would have two separate aggregations

No, there is a single final aggregation processor that performs all the aggregations at the same time. We talked about how there is currently no way to split the computation into separate branches and then join the results. For `SELECT SUM(DISTINCT x), SUM(DISTINCT y)`, your code as written generates a single local distinct processor for each node, plus a single final aggregation processor. Let's look at a plan (`EXPLAIN (DISTSQL) SELECT SUM(DISTINCT x), SUM(DISTINCT y)`.

I will be in the office Mon-Wed, we can go over it more in person.

---

*[pkg/sql/distsql_physical_planner.go, line 850 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgfsT02hu-KsyYHuBlM:-KgfuwnVhq7Nf6Z0t1Fh:b-4skdxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f7780419b953eb2fc769c53857415874d6d2dc/pkg/sql/distsql_physical_planner.go#L850)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

This is almost certainly wrong, and I'm messing up setting it correctly. If you have any thoughts, I'd greatly appreciate it.
</blockquote></details>

We don't need this. The output types of a distinct processor are always the same as the input types, i.e. `p.ResultTypes`.

---

*[pkg/sql/distsql_physical_planner.go, line 859 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgfvMs14n-224riVJoi:-KgfvMs14n-224riVJoj:b-qg5xsg) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f7780419b953eb2fc769c53857415874d6d2dc/pkg/sql/distsql_physical_planner.go#L859)):*
> ```Go
> 		}
> 
> 		if len(p.ResultRouters) > 1 {
> ```

if we have a single result stream, we shouldn't be doing any of this. A single final processor is best in that case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_physical_planner.go, line 802 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgR9zb8UvYyPFnKwNfY:-KggAKQEif9fvs2s17tw:bvyy18o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L802)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

> The first example (SELECT SUM(DISTINCT x), SUM(DISTINCT y)) would have two separate aggregations

No, there is a single final aggregation processor that performs all the aggregations at the same time. We talked about how there is currently no way to split the computation into separate branches and then join the results. For `SELECT SUM(DISTINCT x), SUM(DISTINCT y)`, your code as written generates a single local distinct processor for each node, plus a single final aggregation processor. Let's look at a plan (`EXPLAIN (DISTSQL) SELECT SUM(DISTINCT x), SUM(DISTINCT y)`.

I will be in the office Mon-Wed, we can go over it more in person.
</blockquote></details>

Ah, my mistake, yes this makes sense. You are right that the plan for `explain(distsql) (SELECT sum(Distinct y), sum(distinct z) FROM xyz);` as-is is broken. Here's the plan as is:
https://raduberinde.github.io/decode.html?eJykkTFrwzAQhff-CnNTAxpiZ9PkkiwempQknYIG1TqMwPGZ0xmSBv_3YiuFxoMpBE337r33SegGDTnc2jMG0CdIwShomUoMgXiQoqFwF9BLBb5pOxlko6AkRtA3EC81goaj_apxj9YhgwKHYn09lrbsz5av-eX6DQp2negkz1S-AtMroE5-K3v1f9jGB_FNKY8kYoeMbk110MnJDMu77y4tk-Gk5hn0W1UxVlZo8szD5_vrpjgci-36mOTpAtREyhbPYPcYWmoCPkCnfUYBugrjxwXquMQPpnIsj-NudI-CwyBxm8ahaOJquNbfcDobzubD2Wx4NQmb_uUnAAD__2n42PM

I'm going to try and make some progress here in detecting and just disabling the local distinct processing in these cases.

---

*[pkg/sql/distsql_physical_planner.go, line 850 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgfsT02hu-KsyYHuBlM:-KggAumWbzVZDTO2Bpj1:b-ioxi1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f7780419b953eb2fc769c53857415874d6d2dc/pkg/sql/distsql_physical_planner.go#L850)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We don't need this. The output types of a distinct processor are always the same as the input types, i.e. `p.ResultTypes`.
</blockquote></details>

Done. Thank you.

---

*[pkg/sql/distsql_physical_planner.go, line 859 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgfvMs14n-224riVJoi:-KggAzC_gxNcM_ID5NQT:bgmbm5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f7780419b953eb2fc769c53857415874d6d2dc/pkg/sql/distsql_physical_planner.go#L859)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

if we have a single result stream, we shouldn't be doing any of this. A single final processor is best in that case.
</blockquote></details>

Correct - that's the else branch. We add a single `NoGrouping` stage (since we have a single result stream, so grouping doesn't matter), and then `finalAggSpec` will add an aggregation over that. What am I missing here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsql_physical_planner.go, line 802 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgR9zb8UvYyPFnKwNfY:-KggU1fyRQFrSVj3RmEe:b-n7kjul) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L802)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Ah, my mistake, yes this makes sense. You are right that the plan for `explain(distsql) (SELECT sum(Distinct y), sum(distinct z) FROM xyz);` as-is is broken. Here's the plan as is:
https://raduberinde.github.io/decode.html?eJykkTFrwzAQhff-CnNTAxpiZ9PkkiwempQknYIG1TqMwPGZ0xmSBv_3YiuFxoMpBE337r33SegGDTnc2jMG0CdIwShomUoMgXiQoqFwF9BLBb5pOxlko6AkRtA3EC81goaj_apxj9YhgwKHYn09lrbsz5av-eX6DQp2negkz1S-AtMroE5-K3v1f9jGB_FNKY8kYoeMbk110MnJDMu77y4tk-Gk5hn0W1UxVlZo8szD5_vrpjgci-36mOTpAtREyhbPYPcYWmoCPkCnfUYBugrjxwXquMQPpnIsj-NudI-CwyBxm8ahaOJquNbfcDobzubD2Wx4NQmb_uUnAAD__2n42PM

I'm going to try and make some progress here in detecting and just disabling the local distinct processing in these cases.
</blockquote></details>

What we can do in this case - and what I actually thought the code was already doing, but I may have missed something - is to set up a distinct processor with distinct columns `0, 1`. So we at least eliminate rows that have *both* x and y matching some other row. So for example if the data for x and y is
```
1 1
1 2
1 1
1 3
1 1
2 1
1 2
2 2
2 3
2 2
2 3
```
we at least trim that down to
```
1 1
1 2
1 3
2 1
2 2
2 3
```

Obviously not as good as de-duplicating x and y separately, but may still provide a significant speedup in some cases.

---

*[pkg/sql/distsql_physical_planner.go, line 859 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgfvMs14n-224riVJoi:-KggTjWmzD-DogfVpb8N:b-j47eb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f7780419b953eb2fc769c53857415874d6d2dc/pkg/sql/distsql_physical_planner.go#L859)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Correct - that's the else branch. We add a single `NoGrouping` stage (since we have a single result stream, so grouping doesn't matter), and then `finalAggSpec` will add an aggregation over that. What am I missing here?
</blockquote></details>

I am talking about the entire green part. There's no point in adding a distinct processor at all if we have a single stream.

---

*[pkg/sql/distsql_physical_planner.go, line 828 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KggVySAzjq8bJvWMpd3:-KggVySAzjq8bJvWMpd4:b93ibs8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c198d435ca13e383a167ae86a127450a63158592/pkg/sql/distsql_physical_planner.go#L828)):*
> ```Go
> 	var finalAggSpec distsqlrun.AggregatorSpec
> 
> 	if !multiStage && anyDistinct && allDistinct {
> ```

we don't need `anyDistinct` here (it should always be implied by `allDistinct`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
To add more tests, it's probably best to rebase on top of #14543 (or wait for it to go in), and add some testcases to `distsql_agg`.

Commit:
distsql: add support for distinct in aggregations

arjunravinarayan commented:
PTAL.

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsql_physical_planner.go, line 802 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgR9zb8UvYyPFnKwNfY:-KgzIcM1i70B-Af079YM:bjbs7q4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95332d6b55cbe2066f6bffb38c4c48700e0a0e4/pkg/sql/distsql_physical_planner.go#L802)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What we can do in this case - and what I actually thought the code was already doing, but I may have missed something - is to set up a distinct processor with distinct columns `0, 1`. So we at least eliminate rows that have *both* x and y matching some other row. So for example if the data for x and y is
```
1 1
1 2
1 1
1 3
1 1
2 1
1 2
2 2
2 3
2 2
2 3
```
we at least trim that down to
```
1 1
1 2
1 3
2 1
2 2
2 3
```

Obviously not as good as de-duplicating x and y separately, but may still provide a significant speedup in some cases.
</blockquote></details>

Done. I've added a test in the TestLogic tests that verifies this as well.

---

*[pkg/sql/distsql_physical_planner.go, line 859 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgfvMs14n-224riVJoi:-Kgyi0yMIevFJmOMsGGW:b1z19kz) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f7780419b953eb2fc769c53857415874d6d2dc/pkg/sql/distsql_physical_planner.go#L859)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I am talking about the entire green part. There's no point in adding a distinct processor at all if we have a single stream.
</blockquote></details>

Done. We only add `NoGrouping` stage if we have more than one stream.

---

*[pkg/sql/distsql_physical_planner.go, line 828 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KggVySAzjq8bJvWMpd3:-KgyzbeNIgtu9fRdZMHW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c198d435ca13e383a167ae86a127450a63158592/pkg/sql/distsql_physical_planner.go#L828)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

we don't need `anyDistinct` here (it should always be implied by `allDistinct`).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-KgzuW1C3whidjIdWJ0G:b-83330m)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/flow_diagram.go, line 151 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgzztDfFUmh9lsUJIlL:-KgzztDgpH-U20N0B6KU:bsv8m3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/distsqlrun/flow_diagram.go#L151)):*
> ```Go
> 	details := []string{
> 		fmt.Sprintf("orderedCols: %v", d.OrderedColumns),
> 		fmt.Sprintf("distinctCols: %v", d.DistinctColumns),
> ```

Other processors show columns using the `@1, @2 ..` notation (see colListStr). We can do that here, and we don't need the `distinctCols:` if these are always on the first line. Most consistent with the others would be:
```
Distinct
@1, @2
Ordered: @1
```
Also, let's avoid showing the `Ordered` string if there are no ordered columns (to keep the plans cleaner).

---

*[pkg/sql/testdata/logic_test/distsql_agg, line 97 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh--w55jlGW1j0Fjo14:-Kh--w55jlGW1j0Fjo15:b-5m0w6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/testdata/logic_test/distsql_agg#L97)):*
> ```
> 
> query I
> SELECT DISTINCT (a) FROM data
> ```

I was wondering how this passed without `rowsort`. I believe the answer is that the planNode recursions figure out that `distinct` could use an ordering, and we decide to maintain the primary key ordering (even if there is no `order by`). Please add a comment explaining this. In fact, it would be better to include the plan too and put the comment there.

---

*[pkg/sql/testdata/logic_test/distsql_agg, line 116 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh-0x18vAckbUe7_oh4:-Kh-0x19verV05ZMaXDP:b4ywlf) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/testdata/logic_test/distsql_agg#L116)):*
> ```
> 
> query T
> SELECT URL FROM [EXPLAIN (DISTSQL) SELECT SUM (DISTINCT A) FROM data]
> ```

What I found (from similar non-distsql tests) is that if something changes in planning, it's more useful to see the EXPLAIN failure instead of the query failure (by default the tests show only the first failure), so the EXPLAIN should be first.

---

*[pkg/sql/testdata/logic_test/distsql_agg, line 147 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh--kYHIbQeae_J1-cO:-Kh--kYHIbQeae_J1-cP:b-siu5rx) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/testdata/logic_test/distsql_agg#L147)):*
> ```
> SELECT URL FROM [EXPLAIN (DISTSQL) SELECT DISTINCT a, b FROM data WHERE (a + b + c) = 27 ORDER BY c,b,a]
> ----
> https://raduberinde.github.io/decode.html?eJzMlU9r3DAQxe_9FGJOCTuHSPbmj6AgaCnsJSlpbkEH1xqKwbGMJENL8Hcvtre0axopsN7FF4Mlvfk9PR7oFRpr6L54IQ_yGTggCEDIACEHhC1ohNbZkry3bjgyCXbmJ8grhKppuzAsa4TSOgL5CqEKNYGEp-J7TY9UGHKAYCgUVT1CWle9FO6XMkUoAOFLVQdykl1cKM42TInL4Ztdso9M3Egpd_dPgPDQBckURyVQZaB7BNuFP_Ae32_rm3Vh7khlG1Rig4pvlkR9rnyomjIcwqwz5Mh8srWX7FkPm_tz-6UrxvW7sX9p-7lv30z3Z_TI11mOhK0ly5FAnSp4sc7gE7aWDD6BOlXw2TqDT9haMvgE6lTB5-sMPmFryeATqHO8Q__BPpJvbePpADufpxHI_KDpKfO2cyV9dbYch0-_D-PpccGQD9Mun352zbQ12PpXzKNiEReLqDg7EPO5OIvbvo6j86h6Gxdvo-IE-fqYS99Exbdx8m1UfBcX3x1jmyc6lipZvGU8UTN-VM94omh5Ah5vGk9Ujce7Nveu-w-_AwAA___Ig_iC
> ```

`orderedCols` shouldn't be empty in this plan

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 1331 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh-2Sp8fLvk19WxBhbe:-Kh-2Sp8fLvk19WxBhbf:b-nv4195) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/distsql_physical_planner.go#L1331)):*
> ```Go
> 	var orderedColumns []uint32
> 	for i := 0; i < len(n.columnsInOrder); i++ {
> 		if n.columnsInOrder[i] {
> ```

we need to go through `planToStreamColMap`, like below

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/testdata/logic_test/distsql_agg, line 147 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh--kYHIbQeae_J1-cO:-Kh-4ZQUx9U_OyufKCaX:b-6bbbob) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/testdata/logic_test/distsql_agg#L147)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`orderedCols` shouldn't be empty in this plan
</blockquote></details>

Wait, we are sorting by c,b,a.. so we actually don't have an order on `a` and `b`... I think this is ok.

Fun fact: this query doesn't work with regular sql:
  ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/testdata/logic_test/distsql_agg, line 147 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh--kYHIbQeae_J1-cO:-Kh-5NU1iC2HUCXzZ9qx:b-g6oaks) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/testdata/logic_test/distsql_agg#L147)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Wait, we are sorting by c,b,a.. so we actually don't have an order on `a` and `b`... I think this is ok.

Fun fact: this query doesn't work with regular sql:
  ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list

</blockquote></details>

Sorry, that error was postgres, not regular sql

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsql_physical_planner.go, line 1338 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh-AQ9Pt4wwB7jBwSht:-Kh-AQ9Pt4wwB7jBwShu:baqt5wu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef0b68b0ea37209ba42c14efc8e3d648ecbcba26/pkg/sql/distsql_physical_planner.go#L1338)):*
> ```Go
> 	for i := range n.Columns() {
> 		if plan.planToStreamColMap[i] != -1 {
> 			distinctColumns = append(distinctColumns, uint32(i))
> ```

this should be `planTosTreamColMap[i]` not `i`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
TFTR! All things addressed, PTAL.

---

Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/distsql_physical_planner.go, line 1331 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh-2Sp8fLvk19WxBhbe:-Kh3Zhbfg_Up7ZYv037J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/distsql_physical_planner.go#L1331)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

we need to go through `planToStreamColMap`, like below
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 1338 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh-AQ9Pt4wwB7jBwSht:-Kh3TIWykbO1l11ptHvm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef0b68b0ea37209ba42c14efc8e3d648ecbcba26/pkg/sql/distsql_physical_planner.go#L1338)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

this should be `planTosTreamColMap[i]` not `i`
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/flow_diagram.go, line 151 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-KgzztDfFUmh9lsUJIlL:-Kh3VZVaMoRCLKN6N7om:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/distsqlrun/flow_diagram.go#L151)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Other processors show columns using the `@1, @2 ..` notation (see colListStr). We can do that here, and we don't need the `distinctCols:` if these are always on the first line. Most consistent with the others would be:
```
Distinct
@1, @2
Ordered: @1
```
Also, let's avoid showing the `Ordered` string if there are no ordered columns (to keep the plans cleaner).
</blockquote></details>

Done.

---

*[pkg/sql/testdata/logic_test/distsql_agg, line 97 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh--w55jlGW1j0Fjo14:-Kh3a3wxBNkyOHfGZ11d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/testdata/logic_test/distsql_agg#L97)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I was wondering how this passed without `rowsort`. I believe the answer is that the planNode recursions figure out that `distinct` could use an ordering, and we decide to maintain the primary key ordering (even if there is no `order by`). Please add a comment explaining this. In fact, it would be better to include the plan too and put the comment there.
</blockquote></details>

Done.

---

*[pkg/sql/testdata/logic_test/distsql_agg, line 116 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh-0x18vAckbUe7_oh4:-Kh3VijyTq3iAXPJ8iw4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/testdata/logic_test/distsql_agg#L116)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What I found (from similar non-distsql tests) is that if something changes in planning, it's more useful to see the EXPLAIN failure instead of the query failure (by default the tests show only the first failure), so the EXPLAIN should be first.
</blockquote></details>

Done.

---

*[pkg/sql/testdata/logic_test/distsql_agg, line 147 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh--kYHIbQeae_J1-cO:-Kh3_S6R72pK_UzG-bZd:b-taxed5) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/testdata/logic_test/distsql_agg#L147)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Sorry, that error was postgres, not regular sql
</blockquote></details>

Just to clarify: everything is good here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-Kh3azN6IA-87voPgaLo:b-s5r1j4)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/testdata/logic_test/distsql_agg, line 147 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-Kh--kYHIbQeae_J1-cO:-Kh44gZA1sFXPGI4zont:b1rhzw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/52ff911e7a745b8ab9aa603ab87c4fc217af9a0c/pkg/sql/testdata/logic_test/distsql_agg#L147)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Just to clarify: everything is good here?
</blockquote></details>

Yeah, my bad.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14300#-:-Kh45-K8bwL5_jyW2yQu:b-jbsf8z)*
<!-- Sent from Reviewable.io -->
